### PR TITLE
PWN-6089 - Pick a token screen for new send flow

### PR DESCRIPTION
**Pull request by 1sakov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-6089

## Description of Work

Added new pick a token screen. 
It's hidden behind the debug feature flag for send flow

## Screenshots (Optional)

Screenshot #1 | Screenshot #2 | Screenshot #3
:------------:|:-------------:|:------------:
Screenshot_20221129_171813 | Screenshot_20221129_171824 | Screenshot_20221129_171836

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenFragment.kt [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+private const val ARG_REQUEST_KEY = "ARG_REQUEST_KEY"
+private const val ARG_RESULT_KEY = "ARG_RESULT_KEY"
+
+class NewSelectTokenFragment :
+    BaseMvpFragment<NewSelectTokenContract.View, NewSelectTokenContract.Presenter>(R.layout.fragment_select_token_new),
+    NewSelectTokenContract.View,
+    SearchView.OnQueryTextListener {
+
+    companion object {
+        fun create(tokens: List<Token.Active>, selectedToken: Token.Active?, requestKey: String, resultKey: String) =
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenFragment.kt:39]: selectedToken  -> preSelectedToken?

File: app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenFragment.kt [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    private val buyAnalytics: BuyAnalytics by inject()
+    private val analyticsInteractor: ScreensAnalyticsInteractor by inject()
+
+    private val binding: FragmentSelectTokenNewBinding by viewBinding()
+
+    override val navBarColor: Int = R.color.bg_smoke
+    override val statusBarColor: Int = R.color.bg_smoke
+
+    private val tokenAdapter: NewSelectTokenAdapter by unsafeLazy {
+        NewSelectTokenAdapter {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenFragment.kt:65]: can you pass this lambda as named arg?

File: app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    private val mappedTokens = mutableListOf<SelectTokenItem>()
+
+    override fun load(tokens: List<Token.Active>, selectedToken: Token.Active?) {
+        mapInitialTokens(selectedToken, tokens)
+
+        view?.showTokens(mappedTokens)
+        view?.showEmptyState(isVisible = mappedTokens.isEmpty())
+    }
+
+    override fun search(lowerCasedSearchText: String) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenPresenter.kt:27]: lowerCasedSearchText but maybe add some context that is a query for token?
tokenNameQuery?

File: app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenPresenter.kt [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                    return@mapIndexed SelectableToken(token, ROUNDED)
+                }
+
+                val roundedState = when (index) {
+                    0 -> TOP_ROUNDED
+                    otherTokensSize - 1 -> BOTTOM_ROUNDED
+                    else -> NOT_ROUNDED
+                }
+
+                return@mapIndexed SelectableToken(token, roundedState)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenPresenter.kt:86]: return@ can be removed

File: app/src/main/java/org/p2p/wallet/home/ui/new/adapter/CategoryTitleViewHolder.kt [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import android.view.ViewGroup
+import org.p2p.wallet.databinding.ItemPickTokenCategoryBinding
+import org.p2p.wallet.home.model.SelectTokenItem
+
+class CategoryTitleViewHolder(
+    private val binding: ItemPickTokenCategoryBinding
+) : RecyclerView.ViewHolder(binding.root) {
+
+    constructor(parent: ViewGroup) : this(
+        binding = ItemPickTokenCategoryBinding.inflate(LayoutInflater.from(parent.context), parent, false)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/new/adapter/CategoryTitleViewHolder.kt:14]: can be done as in org.p2p.wallet.send.ui.search.adapter.SearchViewHolder

File: app/src/main/java/org/p2p/wallet/home/ui/new/adapter/NewSelectTokenAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+        if (fields.isNullOrEmpty()) {
+            onBindViewHolder(holder, position)
+            return
+        }
+
+        val item = data[position]
+        fields.forEach { field ->
+            when (field) {
+                DIFF_FIELD_ROUND_STATE -> (holder as NewSelectTokenViewHolder).setCornersRadius(item as SelectableToken)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/new/adapter/NewSelectTokenAdapter.kt:67]: if?

File: app/src/main/res/layout/fragment_select_token_new.xml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:layout_constraintTop_toTopOf="parent"
+        app:menu="@menu/menu_search"
+        app:navigationIcon="@drawable/ic_back"
+        app:title="@string/send_pick_token_title"
+        app:titleTextAppearance="@style/UiKit.TextAppearance.SemiBold.Title3" />
+
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/tokenRecyclerView"
```
Comment (gslevinkov) [inline app/src/main/res/layout/fragment_select_token_new.xml:21]: recyclerViewTokens

File: app/src/main/res/layout/fragment_select_token_new.xml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
+        android:overScrollMode="never"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/toolbar"
+        tools:listitem="@layout/item_pick_token_new" />
+
+    <TextView
+        android:id="@+id/emptyTextView"
```
Comment (gslevinkov) [inline app/src/main/res/layout/fragment_select_token_new.xml:30]: textViewEmptyTitle

File: app/src/main/res/layout/item_pick_token_category.xml [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+<TextView
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:layout_marginHorizontal="16dp"
+    android:paddingVertical="12dp"
+    android:textAppearance="@style/UiKit.TextAppearance.Regular.Text4"
+    android:textColor="@color/text_mountain"
+    tools:text="Chose token" />
```
Comment (gslevinkov) [inline app/src/main/res/layout/item_pick_token_category.xml:11]: Choose token

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenFragment.kt (hunk 1 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenFragment.kt:39]: It doesn't make any difference. 
current naming is full of context and the developer will understand what is this token about. 
Why we need to make the name longer?

File: app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenPresenter.kt (hunk 3 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenPresenter.kt:27]: good one, I like short namings

File: app/src/main/java/org/p2p/wallet/home/ui/new/adapter/NewSelectTokenAdapter.kt (hunk 6 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/home/ui/new/adapter/NewSelectTokenAdapter.kt:67]: I would leave it. 
If we'll need more fields, we'll just add one more raw. 

If we'll use `if`, we'll need to change the block

File: app/src/main/res/layout/item_pick_token_category.xml (hunk 9 above)
Comment (1sakov) [reply app/src/main/res/layout/item_pick_token_category.xml:11]: `Chosen token`

File: app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenFragment.kt (hunk 1 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenFragment.kt:39]: maybe makes sense, yes.
just preSelected gives more context that it was selected before we open SelectToken and we render fragment with such context

Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/home/ui/new/NewSelectTokenFragment.kt:39]: imho that's overhead. 
1. We need to make names shorter to write less
2. We need to name it to allow the other dev to understand what is going on. `selectedToken` looks absolute understandable. 
3. Imho, there is no need to add more context and make the name longer. 
4. That's not a problem, that's not an error, that's not a context absence.  

@kabramovp2p @davran1001 do you think there is additional context needed in naming?

**Pull request merged**